django/db/models/deletion.py

```diff
@@ -340,9 +345,9 @@ def collect(
                         )
                     )
                     sub_objs = sub_objs.only(*tuple(referenced_fields))
-                if sub_objs:
+                if getattr(on_delete, "lazy_sub_objs", False) or sub_objs:
```

Comment:
The whole reasoning for this `lazy_sub_objs` flag is to maintain the previous `Field.on_delete` contract where it would previously only be called if `sub_objs` was not empty but that requires evaluating the queryset which we want to avoid for `SET` and friends as it is necessary.

If we're open to changing the `on_delete` contract or even go through a deprecation period where we warn of the upcoming change and adapt `CASCADE` to return when `not sub_objs` we could do without this logic.

Open to whichever approach seems the best. Maybe we don't worry about changing this contract that much since custom `on_delete` handlers are quite rare anyway?

Reply:
The only issue I can see here is that `UPDATE` is the rows/table locking statement, so we will lock a table even if not necessary :thinking:

Reply:
Are you referring to rows that might be locked by other queries relying on locks (e.g. `select_for_update`)?

Are update statements really locking when they don't match any rows? If they do match rows then they shouldn't be any worse than the previous `IN potentially_long_list_of_ids` queries?

Reply:
As far as I'm aware, on some databases `UPDATE` makes exclusive table lock regardless the set of updated rows :thinking: 
I'm just trying to think about the potential side-effects.

> If they do match rows then they shouldn't be any worse than the previous IN `potentially_long_list_of_ids` queries?

Yes, that's true.

Reply:
To be fair I think your concerns are valid, thanks for raising them, I'm just not sure if this PR makes things worst they were previously.

Given we use this approach for fast-deletes (leaf cascade deletion with signals) it gives me confidence this is an approach worth considering at least. Do these databases are make exclusive locks also when dealing with deletes? I guess we would have learned about them already if it was the case.

Reply:
> Do these databases are make exclusive locks also when dealing with deletes? I guess we would have learned about them already if it was the case.

Yes they do. Good point :100: , agreed, it's worth trying :+1: